Steps and Boilerplate (#2)
vtex-apps/service-course: course/steps/02_events/en.md

> 

Reviewer: Em cima (porque não posso comentar lá :( ) 
Repete: `on this step. For this step`

Reviewer: app that listen -> app that listen**s**

Reviewer: account and workspace where it was fired -> account and workspace where ~it~ they were fired

## Activity

> 1. First, we need to start by starting out the event firing on the `events-example` app. This app will fire an event every X seconds. After runing `vtex link` on its directory, click on the healthcheck route available and a "ok" message should appear on the browser:

Reviewer: ```suggestion
1. First, we are starting the event firing on the `events-example` app. This app will fire an event every X seconds. After running `vtex link` on the `events-example` directory, click on the healthcheck route available and a "ok" message should appear on the browser:
```

   ```diff
   //node/index/ts
   export default new Service<Clients, State, ParamsContext>({
     clients: {
       implementation: Clients,
       options: {
         default: {
           retries: 2,
           timeout: 10000,
         },
   +      events: {
   +        exponentialTimeoutCoefficient: 2,
   +        exponentialBackoffCoefficient: 2,
   +        initialBackoffDelay: 50,
   +        retries: 1,
   +        timeout: TREE_SECONDS_MS,
   +        concurrency: CONCURRENCY,
   +      },
   +    },
   +  },
   })
   ```

   Going by each configuration, we have the following:

   - `exponentialTimeoutCoefficient`: the exponential factor by which the `timeout` will increase in each retry.
   - `exponentialBackoffCoefficient`: the exponential factor by which the `backoff delay` will increase in each retry.
   - `initialBackoffDelay`: the time the app will wait until the next retry.
   - `retries`: the maximum times the app will retry.
   - `timeout`: the timeout until consider a failure attempt.

>    - `concurrency`: the amount of simultaneous processes the event is able to perform.

Reviewer: Tabelinha manêra

   > By adding this code to the `Service`, we are adding to the `Client` of this `Service`, the capability to handle events. At this point, we are not yet using the `Client` itself when handling the event.

> 3. For now, we are only going to create a log when receiving a event. To create this event handler, in the `/node/event` directory, go to the `liveUsersUpdate.ts` file and insert the code below:

Reviewer: ```suggestion
3. For now, we are only going to create a log when receiving an event. To create this event handler, in the `/node/event` directory, go to the `liveUsersUpdate.ts` file and insert the code below:
```

   ```ts
   //node/event/liveUsersUpdate.ts
   export async function updateLiveUsers() {
     console.log('EVENT HANDLER: received event')
   }
   ```

4. Now, we need to declare in the `Service` the reference to this function. On the `/node/index.ts` file, add this code:

   ```diff
   ...
   + import { updateLiveUsers } from './event/liveUsersUpdate'
   ...

   export default new Service<Clients, State, ParamsContext>({
     clients: {
       implementation: Clients,
       options: {
         default: {
           retries: 2,
           timeout: 10000,
         },
         events: {
           exponentialTimeoutCoefficient: 2,
           exponentialBackoffCoefficient: 2,
           initialBackoffDelay: 50,
           retries: 1,
           timeout: TREE_SECONDS_MS,
           concurrency: CONCURRENCY,
         },
       },
     },
   +  events: {
   +    liveUsersUpdate: updateLiveUsers,
   +  },
   })

   ```

> 5. We also need to modify the `service.json` file. In order to listen to events sent, we need to declare this to give the app's service this capability. To do so, replace the code with this one:

Reviewer: ```suggestion
5. We also need to modify the `service.json` file. In order to listen to events sent, we need to declare this to give the app's service this capability. You may do so, by adding the event handling on the `service.json` file:
```

    ![image](https://user-images.githubusercontent.com/43679629/83802091-8c69f380-a680-11ea-82af-a438fb73f40b.png)

    > This healthcheck route access creates a cache context needed for the VTEX IO to fire events. Without it, the `events-example` app won't be able to fire the events our app is going to listen to.

2. We need to add the event handler on the `Service` declaration to refer to what the app is supposed to do when listening to the event. To do so, on the `/node/index.ts` file, add this code on the `Service` declaration:

    ```diff
    //node/index/ts
    export default new Service<Clients, State, ParamsContext>({
      clients: {
        implementation: Clients,
        options: {
          default: {
            retries: 2,
            timeout: 10000,
          },
    +      events: {
    +        exponentialTimeoutCoefficient: 2,
    +        exponentialBackoffCoefficient: 2,
    +        initialBackoffDelay: 50,
    +        retries: 1,
    +        timeout: TREE_SECONDS_MS,
    +        concurrency: CONCURRENCY,
    +      },
    +    },
    +  },
    })
    ```

    Going by each configuration, we have the following:

>     - `exponentialTimeoutCoefficient`: the exponential factor by which the `timeout` will increase in each retry.

Reviewer 2: Do you think it would be better to put the metric in which this arguments should be? Seconds I think 🤔

    - `exponentialBackoffCoefficient`: the exponential factor by which the `backoff delay` will increase in each retry.
    - `initialBackoffDelay`: the time the app will wait until the next retry.
    - `retries`: the maximum times the app will retry.
    - `timeout`: the timeout until consider a failure attempt.
    - `concurrency`: the amount of simultaneous processes the event is able to perform.

    > By adding this code to the `Service`, we are adding to the `Client` of this `Service`, the capability to handle events. At this point, we are not yet using the `Client` itself when handling the event.

3. For now, we are only going to create a log when receiving a event. To create this event handler, in the `/node/event` directory, go to the `liveUsersUpdate.ts` file and insert the code below:

    ```ts
    //node/event/liveUsersUpdate.ts

>     export async function updateLiveUsers() {

Reviewer 3: O que acha de aqui colocar a função já recebendo o `ctx` para, caso o evento tenha dados, mostrar como pode ser recuperado?

Author: ta adicionado nos steps seguintes, pra nao dar o erro de não estar sendo usado

    }
    ```

4. Now, we need to declare in the `Service` the reference to this function. On the `/node/index.ts` file, add this code:

    ```diff
    ...
    + import { updateLiveUsers } from './event/liveUsersUpdate'
    ...

    export default new Service<Clients, State, ParamsContext>({
      clients: {
        implementation: Clients,
        options: {
          default: {
            retries: 2,
            timeout: 10000,
          },
          events: {
            exponentialTimeoutCoefficient: 2,
            exponentialBackoffCoefficient: 2,
            initialBackoffDelay: 50,
            retries: 1,
            timeout: TREE_SECONDS_MS,
            concurrency: CONCURRENCY,
          },
        },
      },
    +  events: {
    +    liveUsersUpdate: updateLiveUsers,
    +  },
    })

    ```

5. We also need to modify the `service.json` file. In order to listen to events sent, we need to declare this to give the app's service this capability. To do so, replace the code with this one:

    ```diff
    //node/service.json
    {
      "memory": 128,
      "ttl": 10,
      "timeout": 10,
      "minReplicas": 2,
      "maxReplicas": 10,
      "workers": 4,

>     +  "events": {

Reviewer 3: Essas mudanças tão ok? Pelo menos no meu Github tá com a impressão de que o diff não tá pegando direito mais
![image](https://user-images.githubusercontent.com/18706156/85634000-d7a86e00-b650-11ea-9022-b122e7a078a8.png)

>     +    "liveUsersUpdate": {

Reviewer 2: Do you think it would be great to add an explanation of how the event system works? Maybe explaining that it has a pubsub architecture?
